streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java

```diff
@@ -53,44 +56,162 @@
+        }
+        while (isEmpty()) {
+            try {
+                log.debug("Detected that the topology is currently empty, going to wait for something to be added");
+                version.topologyCV.await();
+            } catch (final InterruptedException e) {
+                log.debug("StreamThread was interrupted while waiting on empty topology", e);
+            }
+        }
+    }
+
+    public void registerAndBuildNewTopology(final InternalTopologyBuilder newTopologyBuilder) {
+        try {
+            lock();
+            version.topologyVersion.incrementAndGet();
+            log.info("Adding NamedTopology {}, latest topology version is {}", newTopologyBuilder.topologyName(), version.topologyVersion.get());
+            builders.put(newTopologyBuilder.topologyName(), newTopologyBuilder);
+            buildAndVerifyTopology(newTopologyBuilder);
+            version.topologyCV.signalAll();
+        } finally {
+            unlock();
+        }
+    }
+
+    /**
+     * Removes the topology and blocks until all threads on the older version have ack'ed this removal.
+     * IT is guaranteed that no more tasks from this removed topology will be processed
+     */
+    public void unregisterTopology(final String topologyName) {
+        try {
+            lock();
+            version.topologyVersion.incrementAndGet();
+            log.info("Removing NamedTopology {}, latest topology version is {}", topologyName, version.topologyVersion.get());
+            final InternalTopologyBuilder removedBuilder = builders.remove(topologyName);
+            removedBuilder.fullSourceTopicNames().forEach(allInputTopics::remove);
+            removedBuilder.allSourcePatternStrings().forEach(allInputTopics::remove);
+            version.topologyCV.signalAll();
+        } finally {
+            unlock();
+        }
+    }
+
+    public void buildAndRewriteTopology() {
+        applyToEachBuilder(this::buildAndVerifyTopology);
+    }
+
+    private void buildAndVerifyTopology(final InternalTopologyBuilder builder) {
+        builder.rewriteTopology(config);
+        builder.buildTopology();
+
+        // As we go, check each topology for overlap in the set of input topics/patterns
+        final int numInputTopics = allInputTopics.size();
+        final List<String> inputTopics = builder.fullSourceTopicNames();
+        final Collection<String> inputPatterns = builder.allSourcePatternStrings();
+
+        final int numNewInputTopics = inputTopics.size() + inputPatterns.size();
+        allInputTopics.addAll(inputTopics);
+        allInputTopics.addAll(inputPatterns);
+        if (allInputTopics.size() != numInputTopics + numNewInputTopics) {
+            inputTopics.retainAll(allInputTopics);
+            inputPatterns.retainAll(allInputTopics);
+            inputTopics.addAll(inputPatterns);
+            log.error("Tried to add the NamedTopology {} but it had overlap with other input topics: {}", builder.topologyName(), inputTopics);
```

Comment:
Could we print the `inputTopics` and `inputPatterns` as well in the message?

Reply:
You mean print them separately? Sure